Add disabling on `--no-dev`

doekenorg opened the pull request:
Hey @ollieread, 

As promised a PR to be able to disable the behavior.

Unfortunately Composer does not allow adding extra arguments / options to a command. That's why I think it makes sense to listen to `--no-dev` as this plugin is useful for development environments. This is also why I moved event-listener from `init` to `PluginEvents::Command`. This prevents unnecessary extra registers for other commands, as the repositories are only useful during `install` and `update` (works for `i` and `u` as well).

I did some refactoring of the code (PSR-12), hope you don't mind. And I saw they provide a PHPstan file; so I figured I'd add PHPStan and add a helper command (`composer analyse`).

- Resolve composer path via the current instance
- Deduplicate info messages
- Put some messages behind a `verbose` flag
- Moved action to `install` and `update` only to be able to intercept the `--no-dev` flag
- Reading json via JsonFile to avoid errors
- Fixed plugin name in composer.json
- Added PHPStan rules and `analyse` script

doekenorg committed:
Add disabling on `--no-dev`

- Resolve composer path via the current instance
- Deduplicate info messages
- Put some messages behind a `verbose` flag
- Moved action to `install` and `update` only to be able to intercept the `--no-dev` flag
- Reading json via JsonFile to avoid errors
- Fixed plugin name in composer.json
- Added PHPStan rules and `analyse` script

doekenorg committed:
Remove main composer check

This check is no longer needed as only `init` triggered multiple times on an installation.

NickSdot commented:
I also had a chance to play with then plugin, but did run into the below error when installing it. I have tried this PR locally, and can confirm that it solved the issue.

```
  - Installing ollieread/composer-local-repositories (dev-main efd0197): Extracting archive
Plugin initialization failed (Plugin ollieread/composer-local-repositories could not be initialized, class not found: Ollieread\ComposerLocalRepositories\LocalRepositoryPlugin), uninstalling plugin
  - Removing ollieread/composer-local-repositories (dev-main efd0197)
    Install of ollieread/composer-local-repositories failed

In PluginManager.php line 298:
                                                                                              
  Plugin ollieread/composer-local-repositories could not be initialized, class not found: Ol  
  lieread\ComposerLocalRepositories\LocalRepositoryPlugin                                                                                                     
```


### Side Note
Not sure how important this is for you @doekenorg, but regarding not being able to pass a custom flag. You can do something like this:

```bash
composer update -- --foo # note the empty double dash.
```

```php 
var_dump($event->getInput()->getArguments());

// array(2) {
//    ["command"]=>
//    string(6) "update"
//   ["packages"]=>
//   array(1) {
//     [0]=>
//    string(5) "--foo"
//   }
// }
``` 

A bit ugly since it misuses `packages`, but it would work. I am a bit surprised that composer here doesn't work the same as [here](https://github.com/composer/composer/blob/d3d378184bf1ae88abb1c3c9f8d0c26de5ea2744/doc/articles/scripts.md?plain=1#L210). I know update/install/etc allows multiple package names, but that doesn't make the `--` approach impossible, though. However, all this just as a side note. Using the `--no-dev` flag looks good to me.

doekenorg commented:
It can be a configuration key too on the global composer, where you configure on which options it should ignore, or trigger. 

I think it should also inspect the repository for packages, and update those packages to `@dev`. This way the chance of connection is better in case of version mismatches.

doekenorg committed:
Add configuration keys on global composer.json

doekenorg committed:
Force `@dev` on dependencies found in local repositories

doekenorg committed:
Add README

doekenorg commented:
@ollieread For your consideration:

- Added a README with some basic information on how to use the plugin
- Added configuration keys (see README)
- Added a `force-dev` key which change the contraint of a "local package" from whatever it was to `@dev`, and updates the minimum-stability for that package as well. This way you really only have to add a local repository, even if your constriant is something specific like: `^1.1.3`.

Hope you find the time to review this. I can refactor things to be a less 1-file plugin. Make it a bit more SOLID. But I'd like that to be a separate PR. Let me know.

ollieread commented:
Actually, I had time now @doekenorg
